src/System.Runtime.Extensions/src/System/IO/StreamWriter.cs

```diff
@@ -1023,9 +1299,66 @@ private bool HaveWrittenPreamble_Prop
             // Services guys have some perf tests where flushing needlessly hurts.
             if (flushStream)
             {
+                _this.ReturnBuffers();
                 await stream.FlushAsync().ConfigureAwait(false);
             }
         }
+
+        private static async Task FlushAsyncInternalAwaitWrite(Task writeTask, StreamWriter _this, Stream stream)
+        {
+            await writeTask.ConfigureAwait(false);
+
+            _this.ReturnBuffers();
+            await stream.FlushAsync().ConfigureAwait(false);
+        }
+
+        private static Task WritePreambleAsync(StreamWriter _this, Encoding encoding, Stream stream)
+        {
+            _this.HaveWrittenPreamble_Prop = true;
+            byte[] preamble = encoding.GetPreamble();
+            if (preamble.Length > 0)
+            {
+                return stream.WriteAsync(preamble, 0, preamble.Length);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static int EncodeStringBytes(string value, int offset, int length, byte[] byteBuffer, Encoder encoder, bool flushEncoder)
+            => encoder.GetBytes(value.AsReadOnlySpan().Slice(offset, length), new Span<byte>(byteBuffer), flushEncoder);
+
         #endregion
+
+        private char[] CharBuffer
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _rentedCharBuffer ?? RentCharBuffer();
+        }
+
+        private byte[] ByteBuffer
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _rentedByteBuffer ?? RentByteBuffer();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private char[] RentCharBuffer() => _rentedCharBuffer = ArrayPool<char>.Shared.Rent(_charLen);
```

Comment:
@stephentoub @KrzysztofCwalina How much do we care about dropping rented buffers? How far should we go to avoid not returning them?

Reply:
We should strive to return them in all success cases, and in any exceptional cases where it doesn't cost us much to do so.  The worst that happens if we don't return a buffer is we end up paying for collecting a potentially gen2 object that then needs to be replenished, so while it's good to avoid that, if it only happens in exceptional situations, it's not a big deal.

Reply:
@JeremyKuhne do you have a specific scenario where you think this would be hit here?

Reply:
Just writing from multiple threads, not a huge concern just wanted to take the temp on the issue in general.

Reply:
> Just writing from multiple threads

Writing from multiple threads concurrently?  Meaning using the type in a way that's already unsafe?  If so, I'm not concerned about that... many worse things could happen :)

Reply:
Will be in a lot of trouble in general.

I think Console uses `SyncTextWriter` as a wrapper for multiple threads

Reply:
> I'm not concerned about that... 

> Will be in a lot of trouble in general.

I agree. Just looking for bar consensus as we'll be doing a lot of pooling. :)